### PR TITLE
Show warning when annotation type is unknown

### LINKED ISSUE #5070: [ENH] Warn when user provides undeclared symbol in type annotation
### Is your feature request related to a problem? Please describe.

Cython currently ignores annotations with undeclared types - e.g. following code is compiled correctly:

```python
def main():
    a: Bar = 6
```

Compiling code without warning user can lead to errors hard to find - e.g. when developer introduces typo in annotation or forgets to use correct module -e.g.: `a: Py_ssize_t` instead of `a: typing.Py_ssize_t`.

### Describe the solution you'd like.

I would like Cython to raise a warning when annotation contains symbol not defined. When user disables annotation typing with `cython.annotation_typing(False)`, warnings should be silenced.

### Describe alternatives you've considered.

Another possibility is to stop compilation with error but it was rejected as warning is better solution.

### Additional context

See the discussion here: https://github.com/cython/cython/pull/5069#discussion_r989736451

### PR DESCRIPTION
**Pull request by matusvalo**
Fixes #5070

### CODE REVIEW

**Comment by matusvalo** [comment]
PR created for early feedback. Unit tests are pending in this PR.

**Comment by matusvalo** [comment]
> PR created for early feedback. Unit tests are pending in this PR.

If figured out that unittests are enabling all warnings: https://github.com/cython/cython/blob/cf7c067001f19f15c16b4f8608d47822d90ad70c/runtests.py#L2550

So I think there is no need adding unit test for this feature since it is already tested:
https://github.com/cython/cython/blob/cf7c067001f19f15c16b4f8608d47822d90ad70c/tests/run/annotation_typing.pyx#L375

So, I am marking as ready for review.

**Comment by da-woods** [comment]
My feeling is that we should only elevate the error message on the really obvious cases (i.e. name or attribute without an entry). I imagine this might end up being pretty noisy. For example, a regular Python class name is a useful type annotation for the reader but it isn't something that Cython could use.

I haven't really thought how easy it'd be to implement though. My feeling is fairly simple (especially because it doesn't have to be perfect - just identify really easy cases).

I just think we should be a bit careful here so it doesn't end up complaining about everything. Haven't had time to properly think about it though

**Comment by matusvalo** [comment]
> My feeling is that we should only elevate the error message on the really obvious cases (i.e. name or attribute without an entry)

I agree. I double checked that and following example is raising warning:

```python
Bar = list

def main(a: Bar):
    print(a)
```

which should not raise an error in this case.

**Comment by matusvalo** [comment]
Based on feedback marking as draft.

**Comment by matusvalo** [comment]
@da-woods how about this PoC code. For the beginning we can check only following options:
1. whether variable is declared when annotation consist only from variable name. This will warn in following case:
```python
def bar(a: Bar):
    # Bar is not defined.
    pass
```

2. whether annotation contains correct cython module attribute. This will warn in following case:
```python
def bar(a: cython.foo):
    # cython.foo does not exist
    pass
```

I think that this is pretty conservative approach. What do you think?

File: tests/errors/pure_warnings.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+
+def main():
+    foo1: typing.Tuple = None
+    foo1: typing.Bar = None
+    foo2: Bar = 1  # warning
+    foo3: int = 1
+    foo4: cython.int = 1
+    foo5: stdint.bar = 5  # warning
+    foo6: object = 1
```
Comment (matusvalo) [inline tests/errors/pure_warnings.py:16]: Here, I wanted also to test the following case:
```python
foo7: cython.bar = 1
```
But it yields following warnings:
```
14:16: Unknown type declaration 'cython.bar' in annotation, ignoring
25:10: 'cpdef_method' redeclared
36:10: 'cpdef_cname_method' redeclared
977:29: Ambiguous exception value, same as default return value: 0
1018:46: Ambiguous exception value, same as default return value: 0
1108:29: Ambiguous exception value, same as default return value: 0
```

Not sure why there are additional warnings. It is the reason why it is not included in the test...

File: Cython/Compiler/ExprNodes.py [hunk 2]
```diff
@@ -14258,6 +14258,14 @@ def analyse_type_annotation(self, env, assigned_value=None):
         with env.new_c_type_context(in_c_type_context=explicit_ctype):
             arg_type = annotation.analyse_as_type(env)
 
+            if isinstance(annotation, NameNode):
+                if not env.lookup(annotation.name):
+                    warning(annotation.pos, "Unknown type declaration in annotation, ignoring", level=1)
+            if isinstance(annotation, AttributeNode):
+                module_scope = annotation.obj.analyse_as_module(env)
+                if module_scope and module_scope.name == 'cython' and not module_scope.lookup_type(annotation.attribute):
```
[the remark this answers is not in the document]
Comment (matusvalo) [reply Cython/Compiler/ExprNodes.py]: implemented

**Comment by matusvalo** [comment]
Ready for review

**Comment by matusvalo** [comment]
Not sure why tests are failing. Only failure is following:
```
Errors from shard 5:
======================================================================
FAIL: runTest (__main__.EndToEndTest)
[5] End-to-end line_profile_test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/cython/cython/runtests.py", line 1983, in runTest
    self.assertEqual(0, res, "non-zero exit status, last output was:\n%r\n-- stdout:%s\n-- stderr:%s\n" % (
AssertionError: 0 != 1 : non-zero exit status, last output was:
'/opt/hostedtoolcache/Python/3.9.15/x64/bin/python test_profile.py'
-- stdout:Timer unit: 1e-09 s


-- stderr:Traceback (most recent call last):
  File "/home/runner/work/cython/cython/TEST_TMP/5/run/line_profile_test/test_profile.py", line 26, in <module>
    assert_stats(profile, func.__name__)
  File "/home/runner/work/cython/cython/TEST_TMP/5/run/line_profile_test/test_profile.py", line 13, in assert_stats
    assert len(stats.timings) > 0, "No profile stats."
AssertionError: No profile stats.
```

But it does not seems to be connected to this PR.

File: tests/errors/pure_warnings.py (hunk 1 above)
Comment (da-woods) [reply tests/errors/pure_warnings.py:16]: That looks like utility code, but I'm not sure from what.

(Will have a proper look at this PR at some point fairly soon...)

**Comment by da-woods** [comment]
> Hi @da-woods, is there anything pending in this PR?

Mostly just that I've run out of energy a little at the end of the year and need to find time to have a proper look through. It will happen!

**Comment by matusvalo** [comment]
> Mostly just that I've run out of energy a little at the end of the year and need to find time to have a proper look through

No problem take your time and take a rest!

File: Cython/Compiler/ExprNodes.py [hunk 3]
```diff
@@ -14271,8 +14271,34 @@ def analyse_type_annotation(self, env, assigned_value=None):
...
+
         if arg_type is None:
-            warning(annotation.pos, "Unknown type declaration in annotation, ignoring")
+            if isinstance(annotation, NameNode):
+                # Validate annotation in form `var: type`
+                if not env.lookup(annotation.name):
+                    warning(annotation.pos, "Unknown type declaration '%s' in annotation, ignoring" % annotation.name, level=1)
+            elif isinstance(annotation, AttributeNode):
+                # Validate annotation in form `var: module.type`
+                if not env.lookup(annotation.obj.name):
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:14282]: I'm not sure this is necessarily referring to a module. And even if it was, looking up the attribute name in the current scope won't find the name in that module. There's `typing` module support now, so we might want to special case that. And the `.entry` of the attribute's object might also give hints.

File: Cython/Compiler/ExprNodes.py [hunk 4]
```diff
@@ -14271,8 +14271,34 @@ def analyse_type_annotation(self, env, assigned_value=None):
...
+                elif annotation.obj.is_cython_module:
+                    # `module` is cython
+                    module_scope = annotation.obj.analyse_as_module(env)
+                    if module_scope and not module_scope.lookup_type(annotation.attribute):
+                        warning(annotation.pos,
+                                "Unknown type declaration '%s.%s' in annotation, ignoring" % (annotation.obj.name, annotation.attribute),
+                                level=1)
+                else:
+                    module_scope = annotation.obj.analyse_as_module(env)
+                    if module_scope and module_scope.pxd_file_loaded and not annotation.analyse_as_type(env):
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:14296]: We already called `annotation.analyse_as_type()` above, that failed and brought us here.

File: tests/errors/dataclass_e5.pyx [hunk 5]
```diff
@@ -15,7 +15,7 @@ cdef class C:
 _WARNINGS = """
 9:7: Found Python 2.x type 'long' in a Python annotation. Did you mean to use 'cython.long'?
 10:7: Found C type 'Py_ssize_t' in a Python annotation. Did you mean to use a Python type?
```
Comment (scoder) [inline tests/errors/dataclass_e5.pyx:17]: I wonder if this suggestion even makes sense. Users probably meant to use a `cython.*` type here, not a Python type.

File: tests/errors/pure_warnings.py (hunk 1 above)
Comment (scoder) [reply tests/errors/pure_warnings.py:16]: Yes, sadly, we have a couple of utility code files that report warnings. We should eventually make those internal (or, rather, fix them).

File: Cython/Compiler/ExprNodes.py [hunk 6]
```diff
@@ -14271,8 +14271,34 @@ def analyse_type_annotation(self, env, assigned_value=None):
...
+                if not env.lookup(annotation.name):
+                    warning(annotation.pos, "Unknown type declaration '%s' in annotation, ignoring" % annotation.name, level=1)
+            elif isinstance(annotation, AttributeNode):
+                # Validate annotation in form `var: module.type`
+                if not env.lookup(annotation.obj.name):
+                    # `module` is undeclared
+                    warning(annotation.pos,
+                            "Unknown type declaration '%s.%s' in annotation, ignoring" % (annotation.obj.name, annotation.attribute),
+                            level=1)
+                elif annotation.obj.is_cython_module:
```
[the remark this answers is not in the document]
Comment (matusvalo) [reply Cython/Compiler/ExprNodes.py:14286]: Yes, in pure python mode, any annotation that is not recognized in cython is silently ignored and python object is used by default. E.g.:
```python
@cython.cfunc
def bar(a: cython.bar) -> cython.void:
    print(a)
```
Is compiled silently on master branch to:
```c
static void __pyx_f_4test_bar(PyObject *__pyx_v_a) {
```
 The scope of this PR is to  print warning for the cases where we are sure that the annotation is broken. So in this case the PR produces the same C code but raises a warning:

```
 cython -3a test.py
warning: test.py:4:17: Unknown type declaration 'cython.bar' in annotation, ignoring
```
@scoder do you think this check should go elsewhere where this case is handled (and ignored)? My personal opinion is that this warnings should be handled in one single place.

File: Cython/Compiler/ExprNodes.py (hunk 4 above)
Comment (matusvalo) [reply Cython/Compiler/ExprNodes.py]: fixed.

File: tests/errors/dataclass_e5.pyx (hunk 5 above)
Comment (matusvalo) [reply tests/errors/dataclass_e5.pyx:17]: > I wonder if this suggestion even makes sense.
If I understand this comment is connected to this line:
```
10:7: Found C type 'Py_ssize_t' in a Python annotation. Did you mean to use a Python type?
```
Should I fix it in this PR?

File: Cython/Compiler/ExprNodes.py (hunk 3 above)
Comment (matusvalo) [reply Cython/Compiler/ExprNodes.py:14281]: I am pretty new to Cython code-base so maybe here I need a help. This test:

```python
if not env.lookup(annotation.obj.name):
    # `module` is undeclared
    warning(annotation.pos, ...)
```

is checking whether for the annotation expression `A.B`  is `A` name known/defined. If not, it is raising the warning. Is this incorrect? It is pretty the same as previous test:
```
if not env.lookup(annotation.name):
    warning(annotation.pos, "Unknown type declaration '%s' in annotation, ignoring" % annotation.name, level=1)
```

Comment (da-woods) [reply Cython/Compiler/ExprNodes.py:14281]: Not sure if this is quite the point that scoder was making, but

```
annotation.obj.name
```

could fail. `annotation.obj` isn't necessarily a name node. Some (non-exhaustive) examples:

```
mod.a.b  # I think obj will be mod.a
func().b  # obj will be func()  - I can't imagine many times this'll be a useful annotation for Cython, but it shouldn't crash!
(1 + x).b  # again, unlikely
```

There's no need to produce a sensible message for all of these, but they shouldn't crash

File: Cython/Compiler/ExprNodes.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/ExprNodes.py:14286]: I agree with Matus here - I think this is fine (and looks to be tested)

File: tests/errors/dataclass_e5.pyx (hunk 5 above)
Comment (da-woods) [reply tests/errors/dataclass_e5.pyx:17]: I think it'd be in scope to improve it in the PR.

File: tests/errors/pure_warnings.py (hunk 1 above)
Comment (da-woods) [reply tests/errors/pure_warnings.py:16]: So I think what's happening here is that missing an attribute lookup on `cython` causes it to cause `CythonScope.load_cythonscope` which loads "TestClass" and some memoryview code. I think this is harmless (because it does get dropped) but does cause the warnings. Running it outside the test suite gives:

```
warning: TestClass:25:10: 'cpdef_method' redeclared 
warning: TestClass:36:10: 'cpdef_cname_method' redeclared 
warning: View.MemoryView:977:29: Ambiguous exception value, same as default return value: 0
warning: View.MemoryView:1018:46: Ambiguous exception value, same as default return value: 0
warning: View.MemoryView:1108:29: Ambiguous exception value, same as default return value: 0
```

What we've done before: in the `_WARNINGS` string you can have comments and the warnings don't need to be in order. So just add these warnings with a comment at the end of the list of warnings:

```
# Spurious warnings from utility code - not part of the core test
25:10: 'cpdef_method' redeclared
36:10: 'cpdef_cname_method' redeclared  
977:29: Ambiguous exception value, same as default return value: 0
1018:46: Ambiguous exception value, same as default return value: 0
1108:29: Ambiguous exception value, same as default return value: 0
"""
```

File: Cython/Compiler/ExprNodes.py [hunk 7]
```diff
@@ -14272,7 +14272,32 @@ def analyse_type_annotation(self, env, assigned_value=None):
...
+                    module_scope = annotation.obj.analyse_as_module(env)
+                    if module_scope and not module_scope.lookup_type(annotation.attribute):
+                        warning(annotation.pos,
+                                "Unknown type declaration '%s.%s' in annotation, ignoring" % (annotation.obj.name, annotation.attribute),
+                                level=1)
+                else:
+                    module_scope = annotation.obj.analyse_as_module(env)
+                    if module_scope and module_scope.pxd_file_loaded:
+                        warning(annotation.pos,
+                                "Unknown type declaration '%s.%s' in annotation, ignoring" % (annotation.obj.name, annotation.attribute),
```
Comment (da-woods) [inline Cython/Compiler/ExprNodes.py:14297]: The other thing that might help - the `AnnotationNode` stores a string representation of the annotation already so you don't necessarily need to construct the string yourself. Not sure whether that lets you simplify too much might it might be useful

File: tests/errors/pure_warnings.py (hunk 1 above)
Comment (matusvalo) [reply tests/errors/pure_warnings.py:16]: Fixed as proposed by @da-woods.

File: Cython/Compiler/ExprNodes.py (hunk 7 above)
Comment (matusvalo) [reply Cython/Compiler/ExprNodes.py]: Fixed.

File: Cython/Compiler/ExprNodes.py (hunk 3 above)
Comment (matusvalo) [reply Cython/Compiler/ExprNodes.py:14282]: Now I understand. I have limited the validation  only when `annotation.obj.is_name` is true. So for simplicity we won't manage all nasty corner cases. Let me know if you a fine with this solution.

File: tests/errors/dataclass_e5.pyx (hunk 5 above)
Comment (matusvalo) [reply tests/errors/dataclass_e5.pyx]: Fixed to : `Found C type 'Py_ssize_t' in a Python annotation. Did you mean to use a 'cython.Py_ssize_t'?` But @da-woods does this warning have any meaning except unitests? `warning()` call is missing `level=` hence it is raised **only during unittests**:

```
(cython39) matus@MacBook-Pro:~/dev/cython/test$ cat test.py
import cython

def main():
    a: Py_ssize_t = 6
(cython39) matus@MacBook-Pro:~/dev/cython/test$ cython -3 test.py
warning: test.py:4:7: Unknown type declaration 'Py_ssize_t' in annotation, ignoring
```

Comment (da-woods) [reply tests/errors/dataclass_e5.pyx]: Sorry for the slow reply...

Without the `level` call warnings aren't shown by default. But they aren't just for the CI. You can turn them on with compiler options. I think `show_all_warnings` as [a Cythonize argument](https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#cythonize-arguments) will do it.

So it isn't *just* for the unittests, but most people probably don't see them. As I said, it'd be welcome, but not necessary for this PR.

**Comment by da-woods** [comment]
This looks more robust to me so I'm happy with it. I'll click the "approve" button to mark that. I'm tempted to merge this after the next release I think

**Review by da-woods — APPROVED** [review]

File: tests/errors/dataclass_e5.pyx (hunk 5 above)
Comment (matusvalo) [reply tests/errors/dataclass_e5.pyx]: > So it isn't _just_ for the unittests, but most people probably don't see them. As I said, it'd be welcome, but not necessary for this PR.

OK fair enough. I learn new think every day. The message is changed in PR so I am closing this conversation.

**Comment by scoder** [comment]
Given that the next release may end up being the first beta, I'd rather have a change like this included by then.

File: Cython/Compiler/ExprNodes.py [hunk 8]
```diff
@@ -2135,7 +2135,7 @@ def analyse_as_type(self, env):
 
         # Try to give a helpful warning when users write plain C type names.
         if not env.in_c_type_context and PyrexTypes.parse_basic_type(self.name):
-            warning(self.pos, "Found C type '%s' in a Python annotation. Did you mean to use a Python type?" % self.name)
+            warning(self.pos, "Found C type '%s' in a Python annotation. Did you mean to use a 'cython.%s'?" % (self.name, self.name))
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:2138]: Given that we say "Found C type …" and not "Found the C type …", I think we should be consistent and go without the article at the end as well.
```suggestion
            warning(self.pos, "Found C type '%s' in a Python annotation. Did you mean to use 'cython.%s'?" % (self.name, self.name))
```

File: Cython/Compiler/ExprNodes.py [hunk 9]
```diff
@@ -14246,6 +14246,33 @@ def analyse_as_type(self, env):
...
+            if not env.lookup(annotation.obj.name):
+                # `module` is undeclared
+                warning(annotation.pos,
+                        "Unknown type declaration '%s' in annotation, ignoring" % self.string.value, level=1)
+            elif annotation.obj.is_cython_module:
+                # `module` is cython
+                module_scope = annotation.obj.analyse_as_module(env)
+                if module_scope and not module_scope.lookup_type(annotation.attribute):
+                    warning(annotation.pos,
+                            "Unknown type declaration '%s' in annotation, ignoring" % self.string.value, level=1)
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:14267]: Isn't this rather an error than a warning? And, don't we generate an error for this somewhere in `ParseTreeTransforms` already?

**Review by scoder — COMMENTED** [review]
Looks good to me overall. I left some comments.

File: Cython/Compiler/ExprNodes.py (hunk 9 above)
Comment (matusvalo) [reply Cython/Compiler/ExprNodes.py:14267]: > Isn't this rather an error than a warning?

Depends on the overall logic we want to have. Today logic is "Don't error out in pure python mode but rather fallback to python object".

> don't we generate an error for this somewhere in ParseTreeTransforms already?

No, in master following code is compiled without error (in pure python mode):

```
$ cat test.py
import cython

def increment(x: cython.bar):
    return x

print(increment(5))

$ cython test.py
/Users/matus/dev/cython/Cython/Compiler/Main.py:370: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /Users/matus/dev/cython/test/test.py
  tree = Parsing.p_module(s, pxd, full_module_name)
```

**Comment by matusvalo** [comment]
> Given that the next release may end up being the first beta, I'd rather have a change like this included by then.

But I would like not to hold more next beta release with this PR. So please, if this cannot be merged right now don't wait with beta.

File: Cython/Compiler/ExprNodes.py (hunk 9 above)
Comment (scoder) [reply Cython/Compiler/ExprNodes.py:14267]: > Don't error out in pure python mode but rather fallback to python object

However, we do control the `cython.*` namespace. Everything in there should be a valid name or rejected. It doesn't help users to allow arbitrary names (or typos) for `cython.*` names.

Comment (matusvalo) [reply Cython/Compiler/ExprNodes.py:14267]: > However, we do control the cython.* namespace.

Fair enough. I will modify it to error in this case.

Comment (matusvalo) [reply Cython/Compiler/ExprNodes.py]: I have adjusted the PR to raise an error when `cython.*` is not recognized.

File: tests/errors/pure_warnings.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    foo9: mod.a.b
+    foo10: func().b
+    with cython.annotation_typing(False):
+        foo8: Bar = 1
+        foo9: stdint.bar = 5
+        foo10: cython.bar = 1
+
+
+@cython.cfunc
+def bar() -> cython.bar:
```
Comment (matusvalo) [inline tests/errors/pure_warnings.py:28]: Question: I have found out that unrecognised type in return value is yielding an error. So now with this PR we are raising an error + new warning for the same line. Not sure if it is best behaviour.

File: Cython/Compiler/ExprNodes.py [hunk 11]
```diff
@@ -14401,6 +14401,33 @@ def analyse_as_type(self, env):
         # for compatibility when used as a return_type_node, have this interface too
         return self.analyse_type_annotation(env)[1]
 
+    def _warn_on_unknown_annotation(self, env, annotation):
+        """Method checks for cases when user should be warned that annotation contains unkonwn types."""
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:14405]:
```suggestion
        """Method checks for cases when user should be warned that annotation contains unknown types."""
```

File: tests/errors/pure_warnings.py (hunk 10 above)
Comment (scoder) [reply tests/errors/pure_warnings.py:28]: This only applies to C functions, where defining a type probably means that the user really wants that exact type there. An error seems right in that case. Let's keep both for now and remove the warning if it really starts annoying us.

**Review by scoder — COMMENTED** [review]
Looks good overall. Let's get this in.

**Comment by scoder** [comment]
Triggering a new test run.

**Comment by scoder** [comment]
Thanks !
